[libc++][test] Fix issues found by MSVC's STL

StephanTLavavej opened the pull request:
* libcxx/test/support/min_allocator.h
  + Fix `tiny_size_allocator::rebind` which mistakenly said `T` instead of `U`.
* libcxx/test/std/algorithms/alg.modifying.operations/alg.partitions/stable_partition.pass.cpp
  + `std::stable_partition` requires bidirectional iterators.
* libcxx/test/std/containers/sequences/vector.bool/max_size.pass.cpp
  + Fix allocator type given to `std::vector<bool>`. The element types are required to match, [N5008](https://isocpp.org/files/papers/N5008.pdf) \[container.alloc.reqmts\]/5: "*Mandates:* `allocator_type::value_type` is the same as `X::value_type`."
* libcxx/test/std/time/time.clock/time.clock.utc/types.compile.pass.cpp
  + Mark `is_steady` as `[[maybe_unused]]`, as it appears within `LIBCPP_STATIC_ASSERT` only.
* libcxx/test/std/algorithms/alg.modifying.operations/alg.rotate/rotate.pass.cpp
* libcxx/test/std/algorithms/alg.modifying.operations/alg.swap/swap_ranges.pass.cpp
* libcxx/test/std/utilities/utility/utility.swap/swap_array.pass.cpp
  + Fix MSVC warning C4127 "conditional expression is constant". `TEST_STD_AT_LEAST_23_OR_RUNTIME_EVALUATED` was introduced for this purpose, so it should be used consistently.
* libcxx/test/std/numerics/numeric.ops/numeric.ops.gcd/gcd.pass.cpp
  + Fix `gcd()` precondition violation for `signed char`. This test case was causing `-128` to be passed as a `signed char` to `gcd()`, which is forbidden.
* libcxx/test/std/containers/sequences/array/assert.iterators.pass.cpp
* libcxx/test/std/containers/sequences/vector/vector.modifiers/assert.push_back.invalidation.pass.cpp
* libcxx/test/std/input.output/iostream.format/print.fun/no_file_description.pass.cpp
  + Split some REQUIRES and XFAIL lines. This is a "nice to have" for MSVC's internal test harness, which is extremely simple and looks for exact comment matches to skip tests. We can recognize the specific lines "REQUIRES: has-unix-headers" and "XFAIL: msvc", but it's a headache to maintain if they're chained with other conditions.
* libcxx/test/support/sized_allocator.h
  + Fix x86 truncation warnings. `std::allocator` takes `std::size_t`, so we need to `static_cast`.
* libcxx/test/std/input.output/file.streams/fstreams/ifstream.members/offset_range.pass.cpp
  + Fix x86 truncation warning. `std::min()` is returning `std::streamoff`, which was being unnecessarily narrowed to `std::size_t`.
* libcxx/test/std/algorithms/alg.sorting/alg.merge/inplace_merge_comp.pass.cpp
  + Fix MSVC warning C4127 "conditional expression is constant" for an always-true branch. This was very recently introduced by #129008 making `N` constexpr. As it's a local constant just nine lines above, we don't need to test whether 100 is greater than 0.

StephanTLavavej committed:
Fix tiny_size_allocator::rebind.

StephanTLavavej committed:
std::stable_partition requires bidirectional iterators.

StephanTLavavej committed:
Fix allocator type given to std::vector<bool>.

StephanTLavavej committed:
Mark is_steady as [[maybe_unused]].

Fixes error: unused variable 'is_steady' [-Werror,-Wunused-const-variable]

StephanTLavavej committed:
Fix MSVC warning C4127 "conditional expression is constant".

StephanTLavavej committed:
Fix gcd() precondition violation for signed char.

StephanTLavavej committed:
Split some REQUIRES and XFAIL lines.

frederick-vs-ja commented on libcxx/test/std/algorithms/alg.modifying.operations/alg.rotate/rotate.pass.cpp:
```diff
@@ -438,9 +438,8 @@ TEST_CONSTEXPR_CXX20 bool test_vector_bool() {
-#if TEST_STD_VER >= 11
-  if (TEST_STD_VER >= 23 || !TEST_IS_CONSTANT_EVALUATED)
-    types::for_each(types::forward_iterator_list<std::unique_ptr<int>*>(), TestUniquePtr());
+#if TEST_STD_VER >= 11 && TEST_STD_AT_LEAST_23_OR_RUNTIME_EVALUATED
+  types::for_each(types::forward_iterator_list<std::unique_ptr<int>*>(), TestUniquePtr());
```
Oh, IIUC `TEST_STD_AT_LEAST_23_OR_RUNTIME_EVALUATED` is not a preprocess-time condition. (ditto below)

```suggestion
#if TEST_STD_VER >= 11
  if (TEST_STD_AT_LEAST_23_OR_RUNTIME_EVALUATED)
    types::for_each(types::forward_iterator_list<std::unique_ptr<int>*>(), TestUniquePtr());
```

StephanTLavavej replied:
I believe you're right, I'll fix it tomorrow. Thanks!

StephanTLavavej replied:
Fixed, thanks again. Can't believe I forgot what I introduced in #74186!

winner245 commented on libcxx/test/std/containers/sequences/vector.bool/max_size.pass.cpp:
```diff
@@ -71,7 +71,7 @@ TEST_CONSTEXPR_CXX20 bool tests() {
   {
     test(std::vector<bool>());
-    test(std::vector<bool, std::allocator<int> >());
+    test(std::vector<bool, std::allocator<bool> >());
```
I want to share a bit of my thoughts while working on this—please correct me if I am wrong.

I was aware of [[container.alloc.reqmts]/5](https://eel.is/c++draft/container.alloc.reqmts#5): 

> Mandates: allocator_type​::​value_type is the same as X​::​value_type[.](https://eel.is/c++draft/container.alloc.reqmts#5.sentence-1)

If this applies without any exception, we should ensure consistency between the `value_type` of the container and the allocator for `vector<bool, Alloc>`. 

However, there's also the `vector<bool>`-specific note in [[vector.bool.pspc]/2](https://eel.is/c++draft/vector.bool.pspc):

> Unless described below, all operations have the same requirements and semantics as the primary vector template, except that operations dealing with the bool value type map to bit values in the container storage and [allocator_traits​::​construct](https://eel.is/c++draft/allocator.traits.members#lib:allocator_traits,construct) is not used to construct these values[.](https://eel.is/c++draft/vector.bool#pspc-2.sentence-1)

Since `vector<bool, Alloc>` does not use `allocator_traits::construct` to construct its `value_type` (i.e. `bool`), which makes sense given its space-optimized, bit-packed representation, I’m not entirely convinced that enforcing `vector<bool, Alloc>::value_type == allocator_type::value_type` is necessary—or whether this should be considered an exception to [container.alloc.reqmts/5].

For reference, I reviewed several different `vector<bool>` implementations: MSVC-STL enforces this requirement, while libstdc++ and libc++ do not (https://godbolt.org/z/jvc9nnb95).
 
If it turns out that I am wrong, we should consider enforce this within `vector<bool>` in libc++, aligning it with the behavior of other allocator-aware containers. However, I’m still unsure and open to feedback.

frederick-vs-ja replied:
> If it turns out that I am wrong, we should consider enforce this within `vector<bool>` in libc++, aligning it with the behavior of other allocator-aware containers.

I believe such change should belong to another patch.

A more general problem to me is that given `vector<bool>` is not an actual container, so more wording are necessary to specify which requirements of containers are also applied to it.

winner245 replied:
> A more general problem to me is that given vector<bool> is not an actual container, so more wording are necessary to specify which requirements of containers are also applied to it.

I agree that having a clear specification on whether the general container requirements apply to `vector<bool>` would be highly valuable, given the specialty of `vector<bool>`. Specifically, whether [[container.alloc.reqmts]/5](https://eel.is/c++draft/container.alloc.reqmts#5) applies to `vector<bool>` has significant implications: if the answer is yes, then libc++ and libstdc++ would not be conforming and need to be fixed (in a separate patch); if the answer is no, then MSVC-STL would be enforcing stricter requirements than the standard mandates.

StephanTLavavej replied:
Every other `vector<T>` requires that it's given an allocator for `T`, so allowing `vector<bool>` to vary doesn't seem useful to users.

What the container actually does with the allocator is IMO not very important for the requirements placed on users. For example, `list<T>` uses its allocator to allocate an internal node type.

winner245 replied:
I understand `list<T>` internally rebinds the allocator as a node allocator. However, `list<T>` requires that the `value_type` (`T` in this case) elements are constructed using `allocator_traits​::​construct` ([[container.alloc.reqmts]](https://eel.is/c++draft/container.alloc.reqmts#note-2)). This makes it essential to maintain consistency between  `allocator_type::value_type` and `T` for `list<T>` (and other general containers). 

In contrast, the use of `allocator_traits::construct` is explicitly exempted for `vector<bool>`, as stated in [[vector.bool.pspc]/2](https://eel.is/c++draft/vector.bool.pspc): "allocator_traits​::​construct is not used to construct these values." Given this, `vector<bool>` already exhibits behavior that is inconsistent with general container requirements.

StephanTLavavej replied:
Sure, I agree that `allocator_traits::construct` is a difference here, and that `vector<bool>` is weird and inconsistent. I don't think that this creates an exception to the Mandates, though. An LWG issue to clearly decide one way or the other could be useful.

Should I undo this change, which will require me to knownfail this `max_size` test for MSVC's STL?

winner245 replied:
I agree that raising an LWG issue would be a productive way to address this divergence across implementations.

As for the change to this specific test, I am fine with either approach you propose. I don't have a strong preference for one over the other, nor am I certain which would be more appropriate. But once the LWG issue is resolved, we will have a clear path forward.

StephanTLavavej replied:
Thanks! I have submitted an LWG issue via email: "Does `vector<bool, Allocator>` mandate that `Allocator::value_type` is `bool`?"

I would prefer to keep my change to use `allocator<bool>` here. If the LWG issue is resolved in favor of creating an exemption, then MSVC's STL will need to be updated, and I believe that a separate libc++ test to exercise `vector<bool, allocator<Anything>>` should be added (instead of mixing this in to the `max_size` test).

StephanTLavavej replied:
This is [LWG-4228](https://cplusplus.github.io/LWG/issue4228).

philnik777 replied:
I'm definitely in favour of requiring that the allocator type is the same as the container's `value_type`. Anything else is almost certainly a user error. This also allows us to rebind to the `value_type` and make sure that the allocator stays the same, which has found numerous bugs already.

StephanTLavavej committed:
Fix x86 truncation warnings in sized_allocator.h.

sized_allocator.h(40): warning C4244: 'argument': conversion from 'unsigned __int64' to 'const size_t', possible loss of data
sized_allocator.h(43): warning C4244: 'argument': conversion from 'unsigned __int64' to 'const size_t', possible loss of data

std::allocator takes std::size_t, so we need to static_cast.

StephanTLavavej committed:
Fix x86 truncation warning in offset_range.pass.cpp.

offset_range.pass.cpp(49): warning C4244: 'initializing': conversion from 'const _Ty' to 'size_t', possible loss of data

std::min() is returning std::streamoff, which was being unnecessarily narrowed to std::size_t.

StephanTLavavej committed:
Fix MSVC warning C4127 "conditional expression is constant" for an always-true branch.

This was very recently introduced by LLVM-129008 making `N` constexpr. As it's a local constant
just nine lines above, we don't need to test whether 100 is greater than 0.

ldionne commented on libcxx/test/std/containers/sequences/vector/vector.modifiers/assert.push_back.invalidation.pass.cpp:
```diff
@@ -14,7 +14,8 @@
 // including the past-the-end iterator, are invalidated.
 
-// REQUIRES: has-unix-headers, libcpp-has-abi-bounded-iterators-in-vector
+// REQUIRES: has-unix-headers
+// REQUIRES: libcpp-has-abi-bounded-iterators-in-vector
```
FWIW we have a lot of these multiple requirements on a single line, and I don't think we want to commit to removing those. I'm not pushing back against the changes in this patch specifically, but our official stance here is that we're allowed to use `REQUIRES: foo && bar` and `REQUIRES: foo, bar`.

StephanTLavavej replied:
Thanks, understood. Most of the multi-requirements pose no issue for our internal test harness - it's just the ones that cause MSVC skips, like `REQUIRES: has-unix-headers` or `XFAIL: msvc`, and there haven't been a lot of those being introduced or changed. If this becomes more common or disruptive for you, we can add skips on our side, or figure out how to make the internal test harness parser smarter. (We may need to do that for `ADDITIONAL_COMPILE_FLAGS`, which currently it doesn't understand at all.)

ldionne replied:
Let's cross that bridge when/if we get to it, for now these changes are fine.
